### ISSUE TITLE
Determine duration of possibly-VBR MP3 by reading end of file (without needing to load whole file into SampleQueue)

### ISSUE THREAD

**Issue by PaulWoitaschek** [google/ExoPlayer#10487]
### ExoPlayer Version

2.18.1

### Devices that reproduce the issue

Pixel 4 on Android 11, Android 12 Emulator

### Reproducible in the demo app?

Yes

### Reproduction steps

Play the attached media file and seek.

### Expected result

Seeking works like if FLAG_ENABLE_INDEX_SEEKING was disbaled

### Actual result

Instead of seeking, exoplayer will skip to the next chapter.

### Media

Will be sent by mail

### Bug Report

- [X] You will email the zip file produced by `adb bugreport` to dev.exoplayer@gmail.com after filing this issue.

**Comment by icbaker** [comment]
I did the following to reproduce this:
1. Modify the 2.18.1 demo app to enable index seeking for MP3 files by passing the following to the `DefaultMediaSourceFactory` constructor in `PlayerActivity.createMediaSourceFactory()`:
   ```
   new DefaultExtractorsFactory().setMp3ExtractorFlags(Mp3Extractor.FLAG_ENABLE_INDEX_SEEKING))
   ```
2. Play the provided file

I observed:

The demo app seek bar is disabled, and the total duration is shown as 00:00, but the seek-forward-15s and seek-back-5s buttons are enabled. Tapping these seems to work as expected (so seeking works).

I tried playing the file in an un-modified demo app and the progress bar is enabled, the duration is shown and seeking seems to work.

I never observed seeking not working, and I'm not sure what's meant by 'the next chapter' here:
> Instead of seeking, exoplayer will skip to the next chapter.

---

I dug a bit deeper and worked out that the progress bar is disabled because the duration of the content is not known (`player.getContentDuration()` returns `C.TIME_UNSET`). Disabling the progress bar in this case makes sense, because it's impossible to seek a percentage through the content if the duration is not known (which is all the progress bar can really do). I seeked close to the end of the content (to ensure the content is fully read by the player) and then the duration is determined and the progress bar becomes enabled.

So I think the report is imprecise: Seeking works, but the duration is not known. Does that fit with what you observe?

----

Looking into the implementation of `Mp3Extractor` and how it handles the `FLAG_ENABLE_INDEX_SEEKING`, it tries to read the duration from 4 places:
https://github.com/google/ExoPlayer/blob/ab4d37f499ba49bb3cac7938eb03ebf7133ccfe6/library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp3/Mp3Extractor.java#L456-L464

With the `else` block of that code calling through to `getId3TlenUs`, which returns `C.TIME_UNSET` if it doesn't find a `TLEN` ID3 block:
https://github.com/google/ExoPlayer/blob/ab4d37f499ba49bb3cac7938eb03ebf7133ccfe6/library/extractor/src/main/java/com/google/android/exoplayer2/extractor/mp3/Mp3Extractor.java#L588-L600

The file provided doesn't seem to have a `TLEN` ID3 tag. I think this is because it's a constant bitrate file (`mediainfo` thinks it is anyway), so the duration can be inferred just from the bitrate and length of the file. This inference is disabled when using `FLAG_ENABLE_INDEX_SEEKING`, which is why [the documentation recommends only using it on variable bitrate files](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/extractor/mp3/Mp3Extractor.html#FLAG_ENABLE_INDEX_SEEKING):

> Therefore, it should only be used if one of the following is true:
> * The file is small.
> * The bitrate is variable (or it's unknown whether it's variable) and the file does not provide precise enough seeking metadata.

----

In summary:
* Do you agree that the problem is really that the duration of the content is not known, not that seeking doesn't work?
* Since this file is CBR, and therefore byte-offset seeking works - why are you enabling index-based seeking?

**Comment by PaulWoitaschek** [comment]
First of, sorry - my bug report not really detailed and I did not investigate that deep.

> Do you agree that the problem is really that the duration of the content is not known, not that seeking doesn't work?

Yes.

> Since this file is CBR, and therefore byte-offset seeking works - why are you enabling index-based seeking?

I am using that flag in the context of an audiobook player where I do not control the media that is being played.
When users are playing audiobooks, an offset while seeking is disturbing. Therefore I would be fine with a longer playback buffering time with the effect of an accurate seeking.
However as I don't know about the media that is being played I can not really conditionally turn it on and off.

So responding to the following docs:

> The bitrate is variable (or it's unknown whether it's variable) and the file does not provide precise enough seeking metadata.

The latter is true. I don't know if it's variable.

The trouble shooting page says:
https://exoplayer.dev/troubleshooting.html

> In ExoPlayer, we decided to optimize for speed over accuracy in this case and [FLAG_ENABLE_INDEX_SEEKING](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/extractor/mp3/Mp3Extractor.html#FLAG_ENABLE_INDEX_SEEKING) is therefore disabled by default.

This suggests to me that if want accuracy over speed I can safely turn this on.

**Comment by icbaker** [comment]
>> In ExoPlayer, we decided to optimize for speed over accuracy in this case and [FLAG_ENABLE_INDEX_SEEKING](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/extractor/mp3/Mp3Extractor.html#FLAG_ENABLE_INDEX_SEEKING) is therefore disabled by default.
>
> This suggests to me that if want accuracy over speed I can safely turn this on.

I understand. The caveat here is that toggling this flag doesn't change ExoPlayer's loading/buffering behaviour - and the duration can't be determined until we've read the whole file. Note this doesn't affect 'seeking' in the sense of using the `Player.seek...()` methods, it only affects use-cases that require the duration like the progress bar.

So if you want to fully prioritise accuracy over speed, you also need to tell ExoPlayer to read the whole file up-front. I did this hackily for this file by knowing it's about 1h10mins and customising the `LoadControl` passed to `ExoPlayer.Builder.setLoadControl` to ensure we try and buffer 1h30mins (90mins) of content:

```
new DefaultLoadControl.Builder()
    .setBufferDurationsMs(
        /* minBufferMs= */ 90 * 60 * 1000,
        /* maxBufferMs= */ 90 * 60 * 1000,
        DefaultLoadControl.DEFAULT_BUFFER_FOR_PLAYBACK_MS,
        DefaultLoadControl.DEFAULT_BUFFER_FOR_PLAYBACK_AFTER_REBUFFER_MS)
    .setPrioritizeTimeOverSizeThresholds(true)
    .build()
```

In this case, playback takes a little longer to start (even with the file stored locally on the device) but the progress bar works as expected from the beginning of playback.

It's hard to recommend a way that will work for every file, unless you know the upper bound of duration you're going to encounter. It also costs a lot of memory, because you're reading the whole 160MB file into memory.

You're running into common issues of using MP3 for long-form content, and the reason that it's discouraged really (as mentioned in https://github.com/google/ExoPlayer/issues/6787#issuecomment-568180969).

-----

I'm going to turn this issue into an enhancement, tracking whether we can make `Mp3Extractor` read just the end of the file in order to determine the duration (either by default when index seeking is enabled, or behind an additional flag). This may not be possible. I'm also afraid I can't give you an indication on when we will have the time to look into this.

**Comment by ryanheise** [comment]
I have a few thoughts below.

> This may not be possible.

It is surprising that enabling or disabling `FLAG_ENABLE_INDEX_SEEKING` will affect whether ExoPlayer is able to find the duration for the same media file. If it is possible without the flag, then in theory it's possible with the flag, isn't it?

It "seems" like a bug given that one shouldn't expect this flag to affect the duration return value of the API.

> You're running into common issues of using MP3 for long-form content, and the reason that it's discouraged really (as mentioned in https://github.com/google/ExoPlayer/issues/6787#issuecomment-568180969).

There's sort of a big problem with this statement, because the app developers here are not the content creators.

A podcast player app needs to be able to play any audio content that is listed in the iTunes podcast directory, and there is a whole industry out there of podcast hosts that may dictate what audio format they will encode episodes in, as well as self-hosting podcasters who will make their own encoding decisions, although generally MP3 is used. So it is one thing to discourage app developers not to use MP3 for long form content, but we can't control how podcasters encode their podcasts, and it seems to be an established convention to use MP3.

> which is why [the documentation recommends only using it on variable bitrate files](https://exoplayer.dev/doc/reference/com/google/android/exoplayer2/extractor/mp3/Mp3Extractor.html#FLAG_ENABLE_INDEX_SEEKING):

This would be fine if there were a way to detect whether a certain MP3 file uses a variable or constant bit rate. There isn't. So in practice, this directive in the documentation is not actually practical -- nobody in this situation can actually make the determination on whether to enable index seeking or not if the determination is to be based on whether the file is variable/constant bit rate.

So either we would need a way to be able to detect whether the file is variable/constant bit rate, or we just need indexed seeking to work in all cases, even if the file uses constant bit rate. i.e. It's not merely that this would be an improvement or an enhancement, I think actually the current API is unusable in practice (which some might describe as a bug).

Related issue: #9408
